### PR TITLE
Moving functionality tracer_stubs.py

### PR DESCRIPTION
**Pull request by SemyonBevzuk**
Moved functionality from tracer_stubs.py to symbolic.py/py_symbolic. 
Additional decorators (convert_args_to_extractor, convert_args_and_output) have been added to convert arguments.
The tracer_stubs.py has been removed.

### CODE REVIEW

**Comment by SemyonBevzuk** [comment]
@druzhkov-paul

**Comment by druzhkov-paul** [comment]
run tests

File: tools/export.py [hunk 1]
```diff
@@ -211,10 +213,12 @@ def main(args):
         #compression_ctrl.prepare_for_export()
     # END nncf part
 
+    '''
     if args.target == 'openvino' and not args.alt_ssd_export:
         if hasattr(model, 'roi_head'):
             stub_roi_feature_extractor(model.roi_head, 'bbox_roi_extractor')
             stub_roi_feature_extractor(model.roi_head, 'mask_roi_extractor')
+    #'''
```
Comment (druzhkov-paul) [inline tools/export.py]:
```suggestion
```

File: mmdet/utils/deployment/symbolic.py [hunk 2]
```diff
@@ -20,6 +20,49 @@
...
+    def flatten(args):
+        flat_args = []
+        for a in args:
+            if isinstance(a, (list, tuple)):
+                flat_args.extend(flatten(a))
+            else:
+                flat_args.append(a)
+        return flat_args
+
+    def convert_to_cpu(args):
```
Comment (druzhkov-paul) [inline mmdet/utils/deployment/symbolic.py]: Is it still needed?

File: mmdet/utils/deployment/symbolic.py [hunk 3]
```diff
@@ -20,6 +20,49 @@
...
+    def decorator(func):
+        @wraps(func)
+        def wrapped_function(*args):
+            if op_name == 'roi_feature_extractor':
+                # inner, rois, feats
+                ordered_args = args[0], args[2], args[1] 
+                flat_args = flatten(ordered_args)
+                outputs = as_list(func(*flat_args))
+                if len(outputs) == 1:
+                    outputs = outputs[0]
```
Comment (druzhkov-paul) [inline mmdet/utils/deployment/symbolic.py]: What is the purpose to represent outputs as a list?

File: mmdet/utils/deployment/symbolic.py [hunk 4]
```diff
@@ -20,6 +20,49 @@
...
+
+    def as_list(x):
+        if not isinstance(x, (tuple, list)):
+            return [x, ]
+        return x
+    
+    def decorator(func):
+        @wraps(func)
+        def wrapped_function(*args):
+            if op_name == 'roi_feature_extractor':
```
Comment (druzhkov-paul) [inline mmdet/utils/deployment/symbolic.py]: Let's avoid using any custom logic here. Whatever is needed for some particular ad hoc use case, should be moved to a stand alone wrapper/decorator/function.

File: mmdet/models/roi_heads/roi_extractors/single_level.py [hunk 5]
```diff
@@ -5,6 +5,30 @@
 from mmdet.core import force_fp32
 from mmdet.core.utils.misc import topk
 from mmdet.models.builder import ROI_EXTRACTORS
+from mmdet.utils.deployment.symbolic import py_symbolic
+
+from functools import wraps
+
+
+def convert_args_to_extractor(op_name=None):
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/roi_extractors/single_level.py]: Do we need this `op_name` here?

**Review by druzhkov-paul — CHANGES_REQUESTED** [review]

File: mmdet/utils/deployment/symbolic.py (hunk 2 above)
Comment (SemyonBevzuk) [reply mmdet/utils/deployment/symbolic.py]: Fixed

File: mmdet/utils/deployment/symbolic.py (hunk 3 above)
Comment (SemyonBevzuk) [reply mmdet/utils/deployment/symbolic.py]: Fixed

File: mmdet/models/roi_heads/roi_extractors/single_level.py (hunk 5 above)
Comment (SemyonBevzuk) [reply mmdet/models/roi_heads/roi_extractors/single_level.py]: Fixed

File: mmdet/utils/deployment/symbolic.py (hunk 4 above)
Comment (SemyonBevzuk) [reply mmdet/utils/deployment/symbolic.py]: I have moved the logic for handling this case into a function.

File: mmdet/models/roi_heads/roi_extractors/single_level.py [hunk 6]
```diff
@@ -5,6 +5,30 @@
...
+def convert_args_to_extractor():
+    def is_args_correct(args):
+        return isinstance(args[1], (tuple, list)) and isinstance(args[2], torch.Tensor)
+    
+    def repack_args(args):
+        inner = args[0]
+        feats = tuple(args[2:])
+        rois = args[1]
+        ordered_args = inner, feats, rois
+        return ordered_args
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/roi_extractors/single_level.py]:
```suggestion
        self, rois, *feats = args
        return self, feats, rois
```

File: mmdet/utils/deployment/symbolic.py (hunk 4 above)
Comment (druzhkov-paul) [reply mmdet/utils/deployment/symbolic.py]: I still don't think that custom logic should be present here.

**Review by druzhkov-paul — COMMENTED** [review]
A few thoughts on export decorators in general:

* Graph tracing during export puts some constraints on what can be passed in to the autograd Function and symbolic. It might require some changes on the function's interface level. In this case we end up having a kind of a bottleneck from original function's interface to symbolic graph op interface (to make the wrapper transparent for the outside code) and back (to call the function from inside the wrapper). These interface adapters seem to be a generic pattern and are of use only if the function should actually be wrapped. In this regard it looks reasonable to add support for arbitrary adapters to the `py_symbolic`. It'll allow to keep generic logic clear from the custom one, and save a room for customization at the same time.

    ```python
    def py_symbolic(op_name=None, namespace='custom', inside_adapter=None, outside_adapter=None):
        def decorator(func):
            @wraps(func)
            def wrapped_function(*args):

                name = op_name if op_name is not None else func.__name__
                opset = sym_help._export_onnx_opset_version

                if is_registered_op(name, namespace, opset):

                    class XFunction(torch.autograd.Function):
                        @staticmethod
                        def forward(ctx, *xargs):
                            if inside_adapter is not None:
                                xargs = inside_adapter(*xargs)
                            return func(*xargs)

                        @staticmethod
                        def symbolic(g, *xargs):
                            symb = get_registered_op(name, namespace, opset)
                            return symb(g, *xargs)

                    if outside_adapter is not None:
                        args = outside_adapter(*args)
                    return XFunction.apply(*args)
                else:
                    print(args)
                    return func(*args)
            return wrapped_function
        return decorator
    ```

* At the same time inside adapter is probably not needed at all, because original input arguments can be passes to `func` directly, can't they?
    ```python
    def wrapped_function(*args):
        ...
        class XFunction(torch.autograd.Function):
            @staticmethod
            def forward(ctx, *xargs):
                return func(*args)
    ```

* Another thing to mention is keyword-arguments that are not supported by `py_symbolic` at all. But limited support for non-Tensor inputs (== const attributes) looks to be reasonable and not that hard to implement.

* And the final note. As we finalize these changes it'd be good to add a docstring with `py_symbolic` description and its known limitations.

File: mmdet/utils/deployment/symbolic.py [hunk 7]
```diff
@@ -20,7 +20,10 @@
 from torch.onnx.symbolic_registry import register_op, get_registered_op, is_registered_op
 
 
-def py_symbolic(op_name=None, namespace='mmdet_custom'):
+def py_symbolic(op_name=None, namespace='mmdet_custom', outside_adapter=None):
```
Comment (druzhkov-paul) [inline mmdet/utils/deployment/symbolic.py]: As there's only one adapter left, let's simply call it "adapter".

File: mmdet/models/roi_heads/roi_extractors/single_level.py [hunk 8]
```diff
@@ -85,6 +102,7 @@ def roi_rescale(self, rois, scale_factor):
         new_rois = torch.stack((rois[:, 0], x1, y1, x2, y2), dim=-1)
         return new_rois
 
+    @py_symbolic(op_name='roi_feature_extractor', outside_adapter=convert_args_to_extractor)
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/roi_extractors/single_level.py]:
```suggestion
    @py_symbolic(op_name='roi_feature_extractor', outside_adapter=lambda self, feats, rois: (self, rois) + tuple(feats))
```

File: mmdet/utils/deployment/symbolic.py [hunk 9]
```diff
@@ -33,38 +36,22 @@ def wrapped_function(*args):
                 class XFunction(torch.autograd.Function):
                     @staticmethod
                     def forward(ctx, *xargs):
-                        return func(*xargs)
+                        return func(*args)
 
                     @staticmethod
                     def symbolic(g, *xargs):
                         symb = get_registered_op(name, namespace, opset)
                         return symb(g, *xargs)
```
Comment (druzhkov-paul) [inline mmdet/utils/deployment/symbolic.py:77]: What about passing kwargs to symbolic?

File: mmdet/utils/deployment/symbolic.py [hunk 10]
```diff
@@ -257,11 +221,35 @@ def mul(*args, dtype='Long'):
     return out_combined_bboxes
 
 
+def roi_feature_extractor_symbolics(g, inner, rois, *feats):
```
Comment (druzhkov-paul) [inline mmdet/utils/deployment/symbolic.py]: Following the convention that Tensor-data should go first, it'd be great to move `inner` to the last place here.

Another step towards more readable symbolic, would be to extract required attributes explicitly in adapter, rather than delegate it to symbolic itself:
```python
def adapter(self, feats, rois):
    return (rois,) + tuple(feats) + (self.roi_layers[0].out_size[0], self.featmap_strides, self.roi_layers[0].sample_num)

# FIXME. Not sure about default values here.
def symbolic(g, rois, *feats, out_size=7, featmap_strides=(8, 16), sample_num=0):
    ...
```

What do you think? Does it worth it?

File: mmdet/utils/deployment/symbolic.py [hunk 11]
```diff
@@ -20,10 +20,45 @@
...
+            if isinstance(elem, (list, tuple, torch.Tensor)):
+                return True
+        return False
+
+
+def py_symbolic(op_name=None, namespace='mmdet_custom', outside_adapter=None):
+    """
+    The py_symbolic decorator allows you to wrap your custom function in 
+    torch.autograd.Function and associate it with a custom symbolic function
+    to represent the custom operation in a computation graph.
```
Comment (druzhkov-paul) [inline mmdet/utils/deployment/symbolic.py]:
```suggestion text
    The py_symbolic decorator allows associating a function with a custom symbolic function
    that defines its representation in a computational graph.
```

IMO, usage of `torch.autograd.Function` is more related to implementation details.

**Review by druzhkov-paul — COMMENTED** [review]
And please merge the target branch, because there are conflicts now.

File: mmdet/utils/deployment/symbolic.py (hunk 10 above)
Comment (SemyonBevzuk) [reply mmdet/utils/deployment/symbolic.py]: Moved the extraction of arguments to the adapter function.

If we want to transfer tensors first, and after a constant, then there is a problem with an unknown number of tensors, as in the case of `feats`.
To solve this problem, I changed the adapter. It now returns args (tensors) and kwargs (constants): 
```python
def adapter(self, feats, rois):
    return ((rois,) + tuple(feats), 
        {"output_size": self.roi_layers[0].out_size[0],
         "featmap_strides": self.featmap_strides,
         "sample_num": self.roi_layers[0].sample_num})
...
adapter_args, adapter_kwargs = adapter(*args)
```
This complicates the adapter a bit, adds additional requirements to adapter's return values, and adds logic to handle this case in py_symbolic when calling `symbolic`:
```python
@staticmethod
def symbolic(g, *xargs):
    symb = get_registered_op(name, namespace, opset)
    if adapter is not None:
        return symb(g, *xargs, **adapter_kwargs)
    return symb(g, *xargs)
```
Alternatively, we can pass constants to the function first, and then tensors:
```python
def symbolic(g, output_size, featmap_strides, sample_num, rois, *feats):
```

File: mmdet/models/roi_heads/roi_extractors/single_level.py [hunk 12]
```diff
@@ -5,6 +5,14 @@
...
 from mmdet.core.utils.misc import topk
 from mmdet.models.builder import ROI_EXTRACTORS
+from mmdet.utils.deployment.symbolic import py_symbolic
+
+
+def adapter(self, feats, rois):
+    return ((rois,) + tuple(feats), 
+        {"output_size": self.roi_layers[0].out_size[0],
+         "featmap_strides": self.featmap_strides,
+         "sample_num": self.roi_layers[0].sample_num})
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/roi_extractors/single_level.py]: :+1:

Please use single quotes for strings for consistency with the whole codebase.

File: mmdet/utils/deployment/symbolic.py [hunk 13]
```diff
@@ -33,38 +60,25 @@ def wrapped_function(*args):
...
 
                     @staticmethod
                     def symbolic(g, *xargs):
                         symb = get_registered_op(name, namespace, opset)
+                        if adapter is not None:
+                            return symb(g, *xargs, **adapter_kwargs)
                         return symb(g, *xargs)
 
+                if adapter is not None:
+                    adapter_args, adapter_kwargs = adapter(*args)
```
Comment (druzhkov-paul) [inline mmdet/utils/deployment/symbolic.py]: It'd be good to pass `kwargs` to adapter as well.

File: mmdet/models/roi_heads/roi_extractors/single_level.py [hunk 14]
```diff
@@ -84,7 +92,9 @@ def roi_rescale(self, rois, scale_factor):
         y2 = cy + new_h * 0.5
         new_rois = torch.stack((rois[:, 0], x1, y1, x2, y2), dim=-1)
         return new_rois
-
+    
+    # adapter=lambda self, feats, rois: ((rois,) + tuple(feats), {"output_size": self.roi_layers[0].out_size[0], "featmap_strides": self.featmap_strides, "sample_num": self.roi_layers[0].sample_num})
+    @py_symbolic(op_name='roi_feature_extractor', adapter=adapter)
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/roi_extractors/single_level.py]: Lambda becomes too heavy as it handles keyword-argument. So, please, remove the commented code.

File: mmdet/utils/deployment/symbolic.py [hunk 15]
```diff
@@ -20,10 +20,37 @@
...
+
+    A symbolic function cannot receive a collection of tensors as arguments.
+    If your custom function takes a collection of tensors as arguments,
+    then you need to implement an argument converter from the collection 
+    and pass it to the decorator.
+
+    Args:
+        op_name (str): Operation name, must match the registered operation name.
+        namespace (str): Namespace for this operation.
+        adapter (function): Function for converting arguments.
```
Comment (druzhkov-paul) [inline mmdet/utils/deployment/symbolic.py:36]: Please add a few words about how adapter should look like. It should have the same signature as the decorated function and should output `args` and `kwargs` for the symbolic call, right? Anything else?

File: tools/export.py [hunk 16]
```diff
@@ -223,31 +212,30 @@ def main(args):
         cfg.load_from = args.checkpoint
         cfg.resume_from = None
         compression_ctrl, model = wrap_nncf_model(model, cfg, None, get_fake_input)
-        compression_ctrl.prepare_for_export()
+        # TODO: apply the following string for NNCF 1.5.*
+        # compression_ctrl.prepare_for_export()
```
Comment (druzhkov-paul) [inline tools/export.py]: This should not be part of this PR. Seems this is the result of a target branch merge. A bit confused if [7619c5a](https://github.com/openvinotoolkit/mmdetection/pull/84/commits/7619c5a7c56d4fd0eb73afb1e78534d968206c08) is a merge commit or not. It should be a merge.

File: tools/export.py [hunk 17]
```diff
@@ -223,31 +212,30 @@ def main(args):
...
 
     optimize_onnx_graph(onnx_model_path)
 
+    # ?
+    with_text = False
+    if args.target == 'openvino' and not args.alt_ssd_export:
+        if hasattr(model, 'roi_head'):
+            if getattr(model.roi_head, 'with_text', False):
+                with_text = True
+
```
Comment (druzhkov-paul) [inline tools/export.py]: ?

**Review by druzhkov-paul — CHANGES_REQUESTED** [review]

File: tools/export.py (hunk 17 above)
Comment (druzhkov-paul) [reply tools/export.py]: @Ilya-Krylov Old-style tracer stubs have been gone, all custom exporting logic is handled in a unified way now. Please give us a hint at where do you think is a better place for this code? Should we move it to `export_to_onnx` function?

**Comment by druzhkov-paul** [comment]
run tests

File: tools/export.py (hunk 17 above)
Comment (Ilya-Krylov) [reply tools/export.py]: let think about it a bit later, leave it here for a while. I've tested on text-spotting model and it works.

**Review by Ilya-Krylov — APPROVED** [review]

File: tools/export.py [hunk 18]
```diff
@@ -226,28 +215,26 @@ def main(args):
...
         export_to_onnx(model, fake_data, export_name=onnx_model_path, opset=args.opset,
                        alt_ssd_export=getattr(args, 'alt_ssd_export', False),
-                       verbose=True)
+                       target=args.target, verbose=True)
         add_node_names(onnx_model_path)
         print(f'ONNX model has been saved to "{onnx_model_path}"')
 
     optimize_onnx_graph(onnx_model_path)
 
+    # ?
```
Comment (druzhkov-paul) [inline tools/export.py]:
```suggestion
```

**Review by druzhkov-paul — APPROVED** [review]
